plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerContentProvider.java

```diff
@@ -44,197 +41,157 @@
 			}
-			refreshViewer(newValue);
 		}
 	}
 
+	@Override
+	public void dispose() {
+		model.removeListener(listener);
+		model.dispose();
+	}
+
 	/**
-	 * Called to obtain the root elements of the tree viewer,
-	 * which should be Connections
+	 * Called to obtain the root elements of the tree viewer, which should be
+	 * Connections
 	 */
 	@Override
-	public Object[] getExplorerElements(final Object parentElement) {
-		if(parentElement instanceof ConnectionsRegistry){
-			ConnectionsRegistry registry = (ConnectionsRegistry) parentElement;
-			return registry.getAll(Connection.class).toArray();
+	public Object[] getElements(Object inputElement) {
+		if (inputElement instanceof ConnectionsRegistry) {
+			return model.getConnections().toArray();
 		} else {
 			return new Object[0];
 		}
 	}
 	
+	private void handleLoadingException(Object parentElement, Throwable e) {
+		LoadingStub stub = makeStub(parentElement);
+		stub.add(e);
+		asyncExec(()-> refreshViewer(stub));
+	}
+
+	private LoadingStub makeStub(Object parentElement) {
+		LoadingStub stub = stubs.get(parentElement);
+		if (stub == null) {
+			stub= new LoadingStub();
+			stubs.put(parentElement, stub);
+		}
+		return stub;
+	}
+
 	/**
 	 * Called to obtain the children of any element in the tree viewer
 	 */
 	@Override
-	public Object[] getChildrenFor(Object parentElement) {
-		try{
-			if (parentElement instanceof Connection) {
-				Connection connection = (Connection) parentElement;
-				Object[] result = cache.getProjectsFor(connection).toArray();
-				if(result == null || result.length == 0) {
-					result = new Object[]{new NewProjectLinkNode(connection)};
+	public Object[] getChildren(Object parentElement) {
+		if (parentElement instanceof ConnectionWrapper) {
+			ConnectionWrapper connection = (ConnectionWrapper) parentElement;
+			if (connection.load(e -> {
+				handleLoadingException(parentElement, e);
+			})) {
+				return new Object[] { makeStub(parentElement) };
+			} else {
+				stubs.remove(parentElement);
+				Object[] result = connection.getProjects().toArray();
+				if (result == null || result.length == 0) {
+					result = new Object[] { new NewProjectLinkNode((Connection) connection.getConnection()) };
 				}
 				return result;
-			} else if (parentElement instanceof IProjectAdapter) {
-				IProjectAdapter adapter = (IProjectAdapter) parentElement;
-				Collection<Deployment> deployments = new ArrayList<>(adapter.getDeployments());
-				for (Deployment deployment : deployments) {
-					addDeploymentListeners(deployment);
-				}
-				return deployments.toArray();
-			} else if (parentElement instanceof Deployment) {
-				Deployment deployment = (Deployment) parentElement;
-				return handleDeployment(deployment).toArray();
 			}
-		
-		} catch (OpenShiftException e) {
-			addException(parentElement, e);
+		} else if (parentElement instanceof ProjectWrapper) {
+			ProjectWrapper project = (ProjectWrapper) parentElement;
+			if (project.load(e -> {
```

Comment:
Looks like there is an opportunity to refactor some loading logic into a common method with the ConnectionWraper